sender.go

```diff
@@ -147,6 +147,9 @@ func (s *Sender) Send(ctx context.Context, msg *Message, opts ...SendOption) err
 	ctx, span := s.startProducerSpanFromContext(ctx, "sb.Sender.Send")
 	defer span.End()
 
+	s.clientMu.Lock()
```

Comment:
From the discussion in the issue I thought there was a risk of deadlock with this fix.  @catalinaperalta and I had tried acquiring/releasing the lock if `msg.SessionID` is nil (see [here](https://github.com/Azure/azure-service-bus-go/issues/149#issuecomment-557659831)) however in some of our testing it appeared to introduce a bit of a perf regression.  If the perf issue is real I would expect this change to exhibit the same behavior.  Have you observed this?
At any rate I would prefer scoping the lock acquisition/release to the below `if` block unless it's incomplete.

Reply:
Since Go1.14 , the performance of defer has improved significantly.
https://golang.org/doc/go1.14#runtime

Reply:
I don't think the perf had anything to do with defer although we didn't measure that specifically.  Regardless I think it's best to hold the read lock for the shortest duration required.  With the fix as-is, the read lock is held until Send() exits which seems excessive (unless we know it's a requirement for the fix to properly work).

Reply:
Thanks, I got it.
I'll try to fix.

P.S.
I found there is mutex performance improvement in Go1.14.
https://golang.org/doc/go1.14#library